Add lbinomial function based on lbeta

freeboson opened the pull request:
This implementation of `lbinomial` uses the fact that `binomial(n,k) = 1/( (n+1) * B(n - k + 1, k + 1) )`. I have added tests for some simple cases, and then also comparing for very large `n` with `binomial(::BigInt, ::BigInt)`.

The main motivation is evaluating `log(binomial(n,k))` for large `n` with `k` near `n/2` without approximating with Poisson:

```julia
julia> log(binomial(70, 25)) # OK
43.311504703669215

julia> log(binomial(70, 26)) # :(
ERROR: InexactError()
Stacktrace:
 [1] binomial(::Int64, ::Int64) at ./intfuncs.jl:810

julia> log(binomial(BigInt(70), BigInt(26))) # OK, if you can afford it
4.386007065541805521142055198415251198714576775237582336220765505130378175238536e+01

julia> ans - lbinomial(70, 26) # Close enough?
-8.056495743460440321531836165841374176637792344948696218247614638531207882439305e-15
```

In R, this function is available as `lchoose(n, k)` and is a builtin.

freeboson committed:
add lbinomial(n,k)

vtjnash commented on base/special/gamma.jl:
```diff
@@ -155,3 +155,23 @@ Natural logarithm of the absolute value of the [`beta`](@ref)
+Natural logarithm of the [`binomial`](@ref) coefficient.
+"""
+function lbinomial(n::T, k::T) where {T<:Integer}
+    const S = float(T)
```
> ┌ Warning: Deprecated syntax ``const` declaration on local variable` around special/gamma.jl:165.

vtjnash commented on base/special/gamma.jl:
```diff
@@ -155,3 +155,23 @@ Natural logarithm of the absolute value of the [`beta`](@ref)
+"""
+function lbinomial(n::T, k::T) where {T<:Integer}
+    const S = float(T)
+    (k <  0) && return typemin(S)
```
extra space between `<  0`

vtjnash commented on base/special/gamma.jl:
```diff
@@ -155,3 +155,23 @@ Natural logarithm of the absolute value of the [`beta`](@ref)
+    (k == 0 || k == n) && return zero(S)
+    (k == 1) && return log(abs(n))
+    if k > (n>>1)
+       k = n -k
```
space after `-` (`n - k`)

vtjnash commented on base/special/gamma.jl:
```diff
@@ -155,3 +155,23 @@ Natural logarithm of the absolute value of the [`beta`](@ref)
+
+Natural logarithm of the [`binomial`](@ref) coefficient.
+"""
+function lbinomial(n::T, k::T) where {T<:Integer}
```
Also define:
```julia
lbinomial(n::Integer, k::Integer) = lbinomial(promote(n, k)...)
```
?

vtjnash commented on test/math.jl:
```diff
@@ -443,6 +443,17 @@ end
+    @test lbinomial(10,  0) == 0.0
+    @test lbinomial(10, 10) == 0.0
+    @test lbinomial(10,  1)    ≈ log(10)
+    @test lbfixed(200).(0:200) ≈ blbfixed(200).(0:200)
```
maybe add a test for the behavior when `n < 0`?

vtjnash commented on base/special/gamma.jl:
```diff
@@ -155,3 +155,23 @@ Natural logarithm of the absolute value of the [`beta`](@ref)
+    end
+    k > n && return typemin(S)
+    (k == 0 || k == n) && return zero(S)
+    (k == 1) && return log(abs(n))
```
In the code for `binomial`, it looks like this is `log((isodd(k) ? -1 : 1) * n)`

vtjnash replied:
Actually, I guess this is equivalent, it just needs `isodd(k) && throw(DomainError())` in the `n < 0` branch?

freeboson replied:
Well, it's typical for `lbinomial` to actually mean `log(abs(binomial))`. I didn't note that in the docstr however.

freeboson commented:
I've rebased with:
- fixed ws
- tests for `n<0`
- docstring notes `lbinomial(n, k) = log(abs(binomial(n, k)))`
- method that promotes different `n` and `k` of different `Integer` types
- added `lbinomial` to `doc/src/base/math.md`

freeboson commented:
Any interest in `lbinomial(::Real, ::Integer)`, as in R?

dpsanders commented:
Maybe the Combinatorics.jl package would be a good place for this?

StefanKarpinski commented:
Yes, and I think that `lgamma`, `gamma`, `lfact`, `beta` and `lbeta` should all go elsewhere as well.

freeboson commented:
I didn't know where `lbinomial` should go. I saw `lgamma`/`lfact` and `lbeta` all in one place, so it made sense to shove it there. If you are planning on moving those other functions, it makes sense to put `lbinomial` with them.

freeboson commented:
Since #27473 has been merged, I'll just close this and open in `SpecialFunctions.jl`.
